### PR TITLE
feat: add Arbitrum forwarder contracts

### PR DESCRIPTION
**Pull request by bmzig**
This contains L2 forwarder contracts based on the interface https://github.com/across-protocol/contracts/pull/609, under the assumption that the L2 on which it is deployed is Arbitrum. It mostly follows the same format as the Arbitrum_CustomGasToken_Adapter and Arbitrum_Adapter contracts.

### CODE REVIEW

File: contracts/chain-adapters/Arbitrum_CustomGasToken_L2_Forwarder.sol [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+        ArbitrumERC20GatewayLike _l2ERC20GatewayRouter,
+        address _l3RefundL3Address,
+        IERC20 _l2Usdc,
+        ICCTPTokenMessenger _cctpTokenMessenger,
+        FunderInterface _customGasTokenFunder,
+        uint256 _l3MaxSubmissionCost,
+        address _l3SpokePool,
+        address _crossDomainAdmin
+    )
+        CircleCCTPAdapter(_l2Usdc, _cctpTokenMessenger, CircleDomainIds.UNINITIALIZED)
```
Comment (bmzig) [inline contracts/chain-adapters/Arbitrum_CustomGasToken_L2_Forwarder.sol:73]: This is speculation on how CCTP would work on L3s. We may just want to remove this on the forwarders.

**Comment by bmzig** [comment]
There are some refactors I need to do to hopefully cut down a lot on the code duplication, but I think this is the general idea.

File: contracts/chain-adapters/Arbitrum_L2_Forwarder.sol [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+import { CircleCCTPAdapter, CircleDomainIds } from "../libraries/CircleCCTPAdapter.sol";
+
+/**
+ * @notice Contract containing logic to send messages from L2 to Arbitrum-like L3s.
+ * @notice This contract is for interfacing with Arbitrum-like bridges that use the same
+ * native token on L3 as the native token on L2.
+ */
+
+// solhint-disable-next-line contract-name-camelcase
+contract Arbitrum_L2_Forwarder is ArbitrumForwarderInterface, CircleCCTPAdapter {
```
Comment (mrice32) [inline contracts/chain-adapters/Arbitrum_L2_Forwarder.sol:17]: Same question here, but for the regular arbitrum adapter

File: contracts/chain-adapters/Ovm_CustomGasToken_L2_Forwarder.sol [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+/**
+ * @notice Contract containing logic to send messages from an OVM L2 to an Arbitrum-like L3.
+ * @dev This contract is very similar to Arbitrum_CustomGasToken_Adapter. It is meant to bridge
+ * tokens and send messages over a bridge which uses a custom gas token, except this contract makes
+ * the assumption that it is deployed on an OpStack L2.
+ */
+
+// solhint-disable-next-line contract-name-camelcase
+contract Ovm_CustomGasToken_L2_Forwarder is Arbitrum_CustomGasToken_L2_Forwarder {
```
Comment (mrice32) [inline contracts/chain-adapters/Ovm_CustomGasToken_L2_Forwarder.sol:18]: +1, great job on keeping these super simple

**Review by mrice32 — COMMENTED** [review]
Looks great! A few comments

File: contracts/chain-adapters/Arbitrum_Adapter.sol [hunk 4]
```diff
@@ -7,7 +7,8 @@ import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
 import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
 import "../external/interfaces/CCTPInterfaces.sol";
 import "../libraries/CircleCCTPAdapter.sol";
-import { ArbitrumInboxLike as ArbitrumL1InboxLike, ArbitrumL1ERC20GatewayLike } from "../interfaces/ArbitrumBridgeInterfaces.sol";
```
Comment (pxrl) [inline contracts/chain-adapters/Arbitrum_Adapter.sol:10]: This change was only just introduced back in this (not yer merged) PR: https://github.com/across-protocol/contracts/pull/609/files#diff-9986d5b9b57b18f73b5c629869f6d8a6cfdbfa978964dd76de7f4f37d4b5d561

Is it worth backporting the change to that PR, or is the intermediate step necessary?

File: contracts/chain-adapters/Arbitrum_AdapterBase.sol [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+        else {
+            uint256 requiredL1CallValue = _contractHasSufficientEthBalance(RELAY_TOKENS_L2_GAS_LIMIT);
+
+            // Approve the gateway, not the router, to spend the hub pool's balance. The gateway, which is different
+            // per L1 token, will temporarily escrow the tokens to be bridged and pull them from this contract.
+            address erc20Gateway = L1_ERC20_GATEWAY_ROUTER.getGateway(l1Token);
+            IERC20(l1Token).safeIncreaseAllowance(erc20Gateway, amount);
+
+            // `outboundTransfer` expects that the caller includes a bytes message as the last param that includes the
+            // maxSubmissionCost to use when creating an L2 retryable ticket: https://github.com/OffchainLabs/arbitrum/blob/e98d14873dd77513b569771f47b5e05b72402c5e/packages/arb-bridge-peripherals/contracts/tokenbridge/ethereum/gateway/L1GatewayRouter.sol#L232
```
Comment (pxrl) [inline contracts/chain-adapters/Arbitrum_AdapterBase.sol:130]: This seems like a dead link now. I _think_ the correct reference for that contract is: https://github.com/OffchainLabs/arbitrum-classic/blob/551a39b381dcea81e03e7599fcb01fddff4fe96c/packages/arb-bridge-peripherals/contracts/tokenbridge/ethereum/gateway/L1GatewayRouter.sol#L240

We are however referencing their legacy contracts though. They seem to be maintaining compatibility moving forward but I wonder if we should try to reference the current implementations instead.

File: contracts/chain-adapters/Arbitrum_CustomGasToken_L2_Forwarder.sol [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+    }
+
+    function _requireAdminSender() internal virtual override onlyFromCrossDomainAdmin {}
+
+    function _applyL1ToL2Alias(address l1Address) internal pure returns (address l2Address) {
+        // Allows overflows as explained above.
+        unchecked {
+            l2Address = address(uint160(l1Address) + uint160(0x1111000000000000000000000000000000001111));
+        }
+    }
```
Comment (pxrl) [inline contracts/chain-adapters/Arbitrum_CustomGasToken_L2_Forwarder.sol:104]: We have this in the SpokePool as well. Should we factor it out to an "Arbitrum Utils" library for inlining at build time?

File: contracts/chain-adapters/Ovm_L2_Forwarder.sol [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+/**
+ * @notice Contract containing logic to send messages from an OVM L2 to an Arbitrum-like L3.
+ * @dev This contract is very similar to Arbitrum_CustomGasToken_Adapter. It is meant to bridge
+ * tokens and send messages over a bridge which uses a custom gas token, except this contract makes
+ * the assumption that it is deployed on an OpStack L2.
+ */
+
+// solhint-disable-next-line contract-name-camelcase
+contract Ovm_L2_Forwarder is Arbitrum_L2_Forwarder {
```
Comment (pxrl) [inline contracts/chain-adapters/Ovm_L2_Forwarder.sol:19]: I keep getting lost in these naming conventions (i.e. L2 forwarder, l3 adapter) and it gets a bit tricker when chain names are overlaid like this. Is there a concrete naming scheme to be followed? wdyt about building a lexicon that explains how to interpret the names?

**Review by nicholaspai — COMMENTED** [review]
I really like the inheritance tree of this PR but I think the names are clunky. Can we think of a better naming scheme? There are too many underscores for one.

File: contracts/chain-adapters/Arbitrum_Adapter.sol [hunk 8]
```diff
@@ -1,78 +1,59 @@
...
+        uint256 _l2GasPrice
+    )
+        Arbitrum_AdapterBase(
+            _l1ArbitrumInbox,
+            _l1ERC20GatewayRouter,
+            _l2RefundL2Address,
+            _l1Usdc,
+            _cctpTokenMessenger,
+            _circleDomainId,
+            _l2MaxSubmissionCost,
```
Comment (nicholaspai) [inline contracts/chain-adapters/Arbitrum_Adapter.sol:53]: Note for future work: anytime you make changes to existing contract code where its understood that you're refactoring, you should leave a comment saying something like "I changed this code here"

File: contracts/chain-adapters/Arbitrum_CustomGasToken_L2_Forwarder.sol [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import { SafeERC20 } from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import { CircleCCTPAdapter, CircleDomainIds } from "../libraries/CircleCCTPAdapter.sol";
+import { ITokenMessenger as ICCTPTokenMessenger } from "../external/interfaces/CCTPInterfaces.sol";
+import { ArbitrumCustomGasTokenInbox, ArbitrumL1ERC20GatewayLike } from "../interfaces/ArbitrumBridgeInterfaces.sol";
+import { ForwarderBase } from "./ForwarderBase.sol";
+import { Arbitrum_CustomGasToken_AdapterBase, FunderInterface } from "./Arbitrum_CustomGasToken_AdapterBase.sol";
+
+/**
+ * @notice Contract containing logic to send messages from Arbitrum to an AVM L3.
```
Comment (nicholaspai) [inline contracts/chain-adapters/Arbitrum_CustomGasToken_L2_Forwarder.sol:13]: In all contracts, its helpful to note which layer this contract is supposed to be deployed on. In this case, this contract is supposed to be deployed on L2, right?

File: contracts/chain-adapters/Arbitrum_CustomGasToken_L2_Forwarder.sol [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+ * @dev This contract is very similar to Arbitrum_CustomGasToken_Adapter. It is meant to bridge
+ * tokens and send messages over a bridge which uses a custom gas token, except this contract assumes
+ * it is deployed on Arbitrum.
+ * @custom:security-contact bugs@across.to
+ */
+
+// solhint-disable-next-line contract-name-camelcase
+contract Arbitrum_CustomGasToken_L2_Forwarder is Arbitrum_CustomGasToken_AdapterBase, ForwarderBase {
```
Comment (nicholaspai) [inline contracts/chain-adapters/Arbitrum_CustomGasToken_L2_Forwarder.sol:21]: alternative naming option:
- Arbitrum_CustomGasToken_L2Adapter

Makes it more clear this behaves like an Adapter, but its on L1. Also I think the forwarder logic maybe is implied if this extends ForwarderBase

File: contracts/chain-adapters/Arbitrum_CustomGasToken_L2_Forwarder.sol [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+    using SafeERC20 for IERC20;
+
+    modifier onlyFromCrossDomainAdmin() {
+        require(msg.sender == _applyL1ToL2Alias(crossDomainAdmin), "ONLY_CROSS_DOMAIN_ADMIN");
+        _;
+    }
+
+    /**
+     * @notice Constructs new L2 Forwarder.
+     * @dev We normally cannot define a constructor for proxies, but this is an exception since all
```
Comment (nicholaspai) [inline contracts/chain-adapters/Arbitrum_CustomGasToken_L2_Forwarder.sol:31]: I would note here in a @dev note that the cross domain admin will be sit upon initilization of the ForwarderBase, which should be the HubPool

File: contracts/chain-adapters/Arbitrum_CustomGasToken_L2_Forwarder.sol [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+import { CircleCCTPAdapter, CircleDomainIds } from "../libraries/CircleCCTPAdapter.sol";
+import { ITokenMessenger as ICCTPTokenMessenger } from "../external/interfaces/CCTPInterfaces.sol";
+import { ArbitrumCustomGasTokenInbox, ArbitrumL1ERC20GatewayLike } from "../interfaces/ArbitrumBridgeInterfaces.sol";
+import { ForwarderBase } from "./ForwarderBase.sol";
+import { Arbitrum_CustomGasToken_AdapterBase, FunderInterface } from "./Arbitrum_CustomGasToken_AdapterBase.sol";
+
+/**
+ * @notice Contract containing logic to send messages from Arbitrum to an AVM L3.
+ * @dev This contract is very similar to Arbitrum_CustomGasToken_Adapter. It is meant to bridge
+ * tokens and send messages over a bridge which uses a custom gas token, except this contract assumes
```
Comment (nicholaspai) [inline contracts/chain-adapters/Arbitrum_CustomGasToken_L2_Forwarder.sol:15]: I would also note that this contract is designed to be controlled by a cross domain admin which should be the HubPool

File: contracts/chain-adapters/Arbitrum_CustomGasToken_L2_Forwarder.sol [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+     * to send a message via the Inbox successfully, or the message will get stuck.
+     * @param l2Token L2 token to send.
+     * @param amount Amount of L2 tokens to deposit and L3 tokens to receive.
+     */
+    function relayTokens(
+        address l2Token,
+        address,
+        uint256 amount
+    ) external payable override {
+        _relayTokens(l2Token, address(0), amount, l3SpokePool);
```
Comment (nicholaspai) [inline contracts/chain-adapters/Arbitrum_CustomGasToken_L2_Forwarder.sol:82]: you should comment why address(0) can be hardcoded here

File: contracts/chain-adapters/Arbitrum_CustomGasToken_L2_Forwarder.sol (hunk 6 above)
Comment (nicholaspai) [reply contracts/chain-adapters/Arbitrum_CustomGasToken_L2_Forwarder.sol:105]: Same comment as paul. If we add it as an internal library function

File: contracts/chain-adapters/Arbitrum_L2_Forwarder.sol [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+
+/**
+ * @notice Contract containing logic to send messages from L2 to AVM L3s.
+ * @notice This contract is for interfacing with AVM bridges that use the same
+ * native token on L3 as the native token on L2.
+ * @custom:security-contact bugs@across.to
+ */
+
+// solhint-disable-next-line contract-name-camelcase
+contract Arbitrum_L2_Forwarder is ForwarderBase, Arbitrum_AdapterBase {
```
Comment (nicholaspai) [inline contracts/chain-adapters/Arbitrum_L2_Forwarder.sol:20]: alternative name could be `Arbitrum_L2Adapter` for similar reasons to [here](https://github.com/across-protocol/contracts/pull/610/files#r1773668983)

File: contracts/chain-adapters/Ovm_CustomGasToken_L2_Forwarder.sol [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+/**
+ * @notice Contract containing logic to send messages from an OVM L2 to an AVM L3.
+ * @dev This contract is very similar to Arbitrum_CustomGasToken_Adapter. It is meant to bridge
+ * tokens and send messages over a bridge which uses a custom gas token, except this contract makes
+ * the assumption that it is deployed on an OpStack L2.
+ * @custom:security-contact bugs@across.to
+ */
+
+// solhint-disable-next-line contract-name-camelcase
+contract Ovm_CustomGasToken_L2_Forwarder is Arbitrum_CustomGasToken_L2_Forwarder {
```
Comment (nicholaspai) [inline contracts/chain-adapters/Ovm_CustomGasToken_L2_Forwarder.sol:20]: alternative name would be Ovm_CustomGasToken_L2_Adapter

File: contracts/chain-adapters/Ovm_CustomGasToken_L2_Forwarder.sol [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+import { LibOptimismUpgradeable } from "@openzeppelin/contracts-upgradeable/crosschain/optimism/LibOptimismUpgradeable.sol";
+import { Lib_PredeployAddresses } from "@eth-optimism/contracts/libraries/constants/Lib_PredeployAddresses.sol";
+import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import { SafeERC20 } from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+/**
+ * @notice Contract containing logic to send messages from an OVM L2 to an AVM L3.
+ * @dev This contract is very similar to Arbitrum_CustomGasToken_Adapter. It is meant to bridge
+ * tokens and send messages over a bridge which uses a custom gas token, except this contract makes
+ * the assumption that it is deployed on an OpStack L2.
```
Comment (nicholaspai) [inline contracts/chain-adapters/Ovm_CustomGasToken_L2_Forwarder.sol:15]: Can you explain a bit more how this differs from Arbitrum_L2 equivalent? Specifically comment on how the onlyFromCrossDomainAdmin-esque function has changed

File: contracts/chain-adapters/Ovm_L2_Forwarder.sol (hunk 7 above)
Comment (nicholaspai) [reply contracts/chain-adapters/Ovm_L2_Forwarder.sol:20]: I think we should stick with the <chainId>_Adapter_{base} naming scheme:
- chainId makes it clear which chain the contract should be deployed on, or L1 if omitted
- Adapter already is intuitively understood by us as something that sends message to a higher level
- Base contracts are meant to be inherited

File: contracts/chain-adapters/Ovm_CustomGasToken_L2_Forwarder.sol (hunk 16 above)
Comment (bmzig) [reply contracts/chain-adapters/Ovm_CustomGasToken_L2_Forwarder.sol]: https://github.com/across-protocol/contracts/pull/610/commits/05dfc6eefae26f129ab68223a011766894ec144b

File: contracts/chain-adapters/Ovm_CustomGasToken_L2_Forwarder.sol (hunk 15 above)
Comment (bmzig) [reply contracts/chain-adapters/Ovm_CustomGasToken_L2_Forwarder.sol]: https://github.com/across-protocol/contracts/pull/610/commits/05dfc6eefae26f129ab68223a011766894ec144b

File: contracts/chain-adapters/Arbitrum_L2_Forwarder.sol (hunk 14 above)
Comment (bmzig) [reply contracts/chain-adapters/Arbitrum_L2_Forwarder.sol]: https://github.com/across-protocol/contracts/pull/610/commits/05dfc6eefae26f129ab68223a011766894ec144b

File: contracts/chain-adapters/Arbitrum_CustomGasToken_L2_Forwarder.sol (hunk 13 above)
Comment (bmzig) [reply contracts/chain-adapters/Arbitrum_CustomGasToken_L2_Forwarder.sol]: https://github.com/across-protocol/contracts/pull/610/commits/05dfc6eefae26f129ab68223a011766894ec144b

File: contracts/chain-adapters/Arbitrum_CustomGasToken_L2_Forwarder.sol (hunk 12 above)
Comment (bmzig) [reply contracts/chain-adapters/Arbitrum_CustomGasToken_L2_Forwarder.sol]: https://github.com/across-protocol/contracts/pull/610/commits/05dfc6eefae26f129ab68223a011766894ec144b

File: contracts/chain-adapters/Arbitrum_CustomGasToken_L2_Forwarder.sol (hunk 11 above)
Comment (bmzig) [reply contracts/chain-adapters/Arbitrum_CustomGasToken_L2_Forwarder.sol]: https://github.com/across-protocol/contracts/pull/610/commits/05dfc6eefae26f129ab68223a011766894ec144b

File: contracts/chain-adapters/Arbitrum_CustomGasToken_L2_Forwarder.sol (hunk 10 above)
Comment (bmzig) [reply contracts/chain-adapters/Arbitrum_CustomGasToken_L2_Forwarder.sol]: https://github.com/across-protocol/contracts/pull/610/commits/05dfc6eefae26f129ab68223a011766894ec144b

File: contracts/chain-adapters/Arbitrum_CustomGasToken_L2_Forwarder.sol (hunk 9 above)
Comment (bmzig) [reply contracts/chain-adapters/Arbitrum_CustomGasToken_L2_Forwarder.sol]: Yes, it is L2: https://github.com/across-protocol/contracts/pull/610/commits/05dfc6eefae26f129ab68223a011766894ec144b

File: contracts/chain-adapters/Arbitrum_CustomGasToken_L2_Forwarder.sol (hunk 6 above)
Comment (bmzig) [reply contracts/chain-adapters/Arbitrum_CustomGasToken_L2_Forwarder.sol]: Refactored: https://github.com/across-protocol/contracts/pull/610/commits/05dfc6eefae26f129ab68223a011766894ec144b

File: contracts/chain-adapters/Arbitrum_AdapterBase.sol (hunk 5 above)
Comment (bmzig) [reply contracts/chain-adapters/Arbitrum_AdapterBase.sol]: Nice catch https://github.com/across-protocol/contracts/pull/610/commits/05dfc6eefae26f129ab68223a011766894ec144b

File: contracts/chain-adapters/Arbitrum_CustomGasToken_L2_Forwarder.sol [hunk 17]
```diff
@@ -0,0 +1,184 @@
...
+
+/**
+ * @notice Contract containing logic to send messages from Arbitrum to an Arbitrum-like L3.
+ * @dev This contract is very similar to Arbitrum_CustomGasToken_Adapter. It is meant to bridge
+ * tokens and send messages over a bridge which uses a custom gas token, except this contract assumes
+ * it is deployed on Arbitrum.
+ */
+
+// solhint-disable-next-line contract-name-camelcase
+contract Arbitrum_CustomGasToken_L2_Forwarder is ArbitrumForwarderInterface, CircleCCTPAdapter {
```
[the remark this answers is not in the document]
Comment (bmzig) [reply contracts/chain-adapters/Arbitrum_CustomGasToken_L2_Forwarder.sol]: There was a big refactor here https://github.com/across-protocol/contracts/pull/610/commits/ea2b6ca3f2a640b1f748c1dbd1324adc5bd10eab which should share code via inheritance.

**Comment by bmzig** [comment]
Closing in favor of https://github.com/across-protocol/contracts/pull/625 since we no longer wish to inherit specific adapter code, but instead delegatecall external contracts which contain bridging logic.

**Pull request closed without merging**